sdk/core/core-lro/src/lroPoller/stateMachine.ts

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { processAzureAsyncOperationResult } from "./azureAsyncPolling";
+import { isBodyPollingDone, processBodyPollingOperationResult } from "./bodyPolling";
+import { processLocationPollingOperationResult } from "./locationPolling";
+import {
+  FinalStateVia,
+  LRO,
+  LROConfig,
+  LROStatus,
+  PollerConfig,
+  RawResponse,
+  ResumablePollOperationState
+} from "./models";
+import { processPassthroughOperationResult } from "./passthrough";
+import { getPollingURL, inferLROMode } from "./requestUtils";
+
+/**
+ * creates a stepping function that maps an LRO state to another.
+ */
+export function createGetLROStatusFromResponse<TResult>(
+  lroPrimitives: LRO<TResult>,
+  config: LROConfig,
+  finalStateVia?: FinalStateVia
+): (rawResponse: RawResponse, flatResponse: TResult) => LROStatus<TResult> {
+  switch (config.mode) {
+    case "AzureAsync": {
+      return processAzureAsyncOperationResult(
+        lroPrimitives,
+        config.resourceLocation,
+        finalStateVia
+      );
+    }
+    case "Location": {
+      return processLocationPollingOperationResult;
+    }
+    case "Body": {
+      return processBodyPollingOperationResult;
+    }
+    default: {
+      return processPassthroughOperationResult;
+    }
+  }
+}
+
+/**
+ * Creates a polling operation that returns a LRO state.
+ */
+export function createPollForLROStatus<TResult>(
+  lroPrimitives: LRO<TResult>,
+  config: LROConfig
+): (pollingURL: string, pollerConfig: PollerConfig) => Promise<LROStatus<TResult>> {
+  return async (path: string, pollerConfig: PollerConfig): Promise<LROStatus<TResult>> => {
+    const response = await lroPrimitives.sendPollRequest(config, path);
+    const retryAfter: string | undefined = response.rawResponse.headers["Retry-After"];
+    if (retryAfter !== undefined) {
+      const retryAfterInMs = parseInt(retryAfter);
+      pollerConfig.intervalInMs = isNaN(retryAfterInMs)
+        ? calculatePollingIntervalFromDate(new Date(retryAfter), pollerConfig.intervalInMs)
+        : retryAfterInMs;
+    }
+    return response!;
```

Comment:
Why do we need the `!` here? I thought `sendPollRequest` always returns a value

Reply:
You're right. It fell through the cracks when I was doing the decoupling from core-client. Before, it was being set inside the callback.

Reply:
Removed `!`.